indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java

```diff
@@ -0,0 +1,270 @@
+    supervisorTaskChecker.scheduleAtFixedRate(
+        () -> {
+          for (StorageLocation location : intermediarySegmentsLocations) {
+            final File[] dirsPerSupervisorTask = location.getPath().listFiles();
+            if (dirsPerSupervisorTask != null) {
+              for (File supervisorTaskDir : dirsPerSupervisorTask) {
+                final String supervisorTaskId = supervisorTaskDir.getName();
+                supervisorTaskCheckTimes.computeIfAbsent(
+                    supervisorTaskId,
+                    k -> DateTimes.nowUtc().plus(intermediaryPartitionTimeout)
+                );
+              }
+            }
+          }
+        },
+        intermediaryPartitionDiscoveryPeriodSec,
+        intermediaryPartitionDiscoveryPeriodSec,
+        TimeUnit.SECONDS
+    );
+    // Check supervisorTask status if its partitions have not been accessed in timeout.
+    // Delete partitions if the supervisorTask is already finished.
+    // Note that the overlord sends a cleanup request when a supervisorTask is finished. The below check is to trigger
+    // the self-cleanup for when middleManager misses the cleanup request from the overlord.
+    supervisorTaskChecker.scheduleAtFixedRate(
+        () -> {
+          final DateTime now = DateTimes.nowUtc();
+          final Set<String> expiredSupervisorTasks = new HashSet<>();
+          for (Entry<String, DateTime> entry : supervisorTaskCheckTimes.entrySet()) {
+            final String supervisorTaskId = entry.getKey();
+            final DateTime checkTime = entry.getValue();
+            if (checkTime.isAfter(now)) {
+              expiredSupervisorTasks.add(supervisorTaskId);
+            }
+          }
+
+          final Map<String, TaskStatus> taskStatuses = indexingServiceClient.getTaskStatuses(expiredSupervisorTasks);
+          RuntimeException exception = null;
+          for (Entry<String, TaskStatus> entry : taskStatuses.entrySet()) {
+            final String supervisorTaskId = entry.getKey();
+            final TaskStatus status = entry.getValue();
+            if (status.getStatusCode().isComplete()) {
+              try {
+                deletePartitions(supervisorTaskId);
+              }
+              catch (IOException e) {
+                if (exception == null) {
+                  exception = new RuntimeException(e);
+                } else {
+                  exception.addSuppressed(e);
+                }
+              }
+            }
+          }
+
+          if (exception != null) {
+            throw exception;
+          }
+        },
+        intermediaryPartitionCleanupPeriodSec,
+        intermediaryPartitionCleanupPeriodSec,
+        TimeUnit.SECONDS
+    );
+  }
+
+  @LifecycleStop
+  public void stop() throws InterruptedException
+  {
+    if (supervisorTaskChecker != null) {
+      supervisorTaskChecker.shutdownNow();
+      supervisorTaskChecker.awaitTermination(10, TimeUnit.SECONDS);
+      supervisorTaskChecker = null;
```

Comment:
what is the use of setting it null here ?

Reply:
it piqued my interest as usually you would do this sort of thing if you wanted to make something eligible for GC if it was a big object and part of large method so as to let it be GC'd asap or if you wanted to do something along the lines of making `IntermediaryDataManager` restartable etc.
I was trying to understand why you did that, but sounds like there is no reason so better to remove it :)

Reply:
Removed.